### PR TITLE
feat: add replicate through time module

### PR DESCRIPTION
**Pull request by cpehle**
This feature will implement a general mechanism for replicating layers through time, in order to implement Batch Norm Through Time (BNTT) (#345).

### CODE REVIEW

File: norse/torch/module/replicate_through_time.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import torch
+
+class ReplicateThroughTime(torch.nn.Module):
+    def __init__(self, module, T):
```
Comment (Jegp) [inline norse/torch/module/replicate_through_time.py:4]: A brief docstring would be helpful

Comment (Jegp) [reply norse/torch/module/replicate_through_time.py:4]: Also, types would be neat :)

File: norse/torch/module/replicate_through_time.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        self.module_list = torch.nn.ModuleList([module for _ in range(T)])
+        self.T = T
+
+    def forward(self, x):
+        outputs = []
+        for ts in range(self.T):
+            y = self.module_list[ts](x[ts])
+            outputs.append(y)
+
+        return torch.stack(outputs)
```
Comment (Jegp) [inline norse/torch/module/replicate_through_time.py:14]: This will break if the module returns tuples (output, state).

**Review by Jegp — CHANGES_REQUESTED** [review]
Nice addition!

I'm mostly concerned with how the `stack` will break if the module is recurrent (returns a tuple of output and state). A quick fix would be to distinguish on the return type. If you care about performance, you could do the check in the constructor.

Also, it would be nice to add things like docstrings and typing :-)

File: norse/torch/module/replicate_through_time.py (hunk 2 above)
Comment (cpehle) [reply norse/torch/module/replicate_through_time.py:14]: Yes, but this isn't meant for tuples.

Comment (cpehle) [reply norse/torch/module/replicate_through_time.py:14]: Rather this is meant for modules which are applied "point wise" in time but with one module per timepoint

**Comment by Jegp** [comment]
I see your point about not wanting this for stateful modules. I'd say let's go ahead. If you can add a few lines of documentation, that would be wonderfully helpful.

**Comment by Jegp** [comment]
Closed for now. Contributions are not than welcome! ❤️
